File: Problems Solutions/155. Min Stack.py
# https://leetcode.com/problems/min-stack/
from utils import display_func_value

class MinStack:

    def __init__(self):
        self.stack = []
        self.mins = []        

    def push(self, val: int) -> None:
        self.stack.append(val)

        if not self.mins:
            self.mins.append(val)
        else:
            self.mins.append(min(val, self.mins[-1]))

    def pop(self) -> None:
        self.stack.pop()
        self.mins.pop()

    @display_func_value
    def top(self) -> int:
        return self.stack[-1]
    
    @display_func_value
    def getMin(self) -> int:
        return self.mins[-1]

minStack = MinStack()
minStack.push(-2)
minStack.push(0)
minStack.push(-3)
minStack.getMin() # return -3
minStack.pop()
minStack.top()    # return 0
minStack.getMin() # return -2

File: Problems Solutions/go/155. Min Stack.go
package main

type MinStack struct {
    values []int
    mins []int
}


func getMin(a, b int) int {
    if a > b {
        return b
    } else {
        return a
    }
}


func CreateMinStack() MinStack {
    return MinStack{}
}


func (this *MinStack) Push(val int)  {
    this.values = append(this.values, val)

    if len(this.mins) == 0 {
        this.mins = append(this.mins, val)
    } else {
        this.mins = append(
            this.mins, 
            getMin(
                this.mins[len(this.mins) - 1], 
                this.values[len(this.values) - 1],
            ),
        )
    }
}


func (this *MinStack) Pop()  {
    this.values = this.values[:len(this.values) - 1]
    this.mins = this.mins[:len(this.mins) - 1]
}


func (this *MinStack) Top() int {
    return this.values[len(this.values) - 1]
}


func (this *MinStack) GetMin() int {
    return this.mins[len(this.mins) - 1]
}